diff --git a/xonsh/main.py b/xonsh/main.py
index 74f2f98..c9320cd 100644
--- a/xonsh/main.py
+++ b/xonsh/main.py
@@ -20,12 +20,15 @@ from xonsh.parser import Parser
 from xonsh.shell import Shell
 from xonsh.timings import setup_timings
-from xonsh.tools import (
-    display_error_message,
-    print_color,
-    print_exception,
-    to_bool_or_int,
-    unquote,
-)
+from xonsh.lazy_imports import lazy_import_module
+tools = lazy_import_module("xonsh.tools")
+display_error_message = tools.display_error_message
+print_color = tools.print_color
+print_exception = tools.print_exception
+to_bool_or_int = tools.to_bool_or_int
+unquote = tools.unquote
 from xonsh.xonfig import print_welcome_screen
 from xonsh.xontribs import auto_load_xontribs_from_entrypoints, xontribs_load
 
diff --git a/xonsh/built_ins.py b/xonsh/built_ins.py
index 1827342..d45041a 100644
--- a/xonsh/built_ins.py
+++ b/xonsh/built_ins.py
@@ -18,7 +18,8 @@ import threading
 import builtins
 from collections.abc import MutableMapping, Sequence, MutableSequence, MutableSet
 
-from xonsh.completers.completer import Completer
+from xonsh.lazy_imports import lazy_import_object
+Completer = lazy_import_object('xonsh.completers.completer', 'Completer')
 from xonsh.dirstack import DirectoryStack, DIRSTACK
 from xonsh.environ import Environ
 from xonsh.aliases import Aliases
diff --git a/xonsh/completers/__init__.py b/xonsh/completers/__init__.py
index a24cb01..f8d7f4c 100644
--- a/xonsh/completers/__init__.py
+++ b/xonsh/completers/__init__.py
@@ -1,8 +1,10 @@
 """Completers for xonsh."""
 
-from xonsh.completers.bash_completion import bash_complete_line
-from xonsh.completers.python import python_signature_complete
-from xonsh.completers.path import complete_path
+from xonsh.lazy_imports import lazy_import_object
+bash_complete_line = lazy_import_object('xonsh.completers.bash_completion', 'bash_complete_line')
+python_signature_complete = lazy_import_object('xonsh.completers.python', 'python_signature_complete')
+complete_path = lazy_import_object('xonsh.completers.path', 'complete_path')
 
 __all__ = [
     'bash_complete_line',
